machine/v1/types_alibabaprovider.go

```diff
@@ -0,0 +1,340 @@
+)
+
+// AlibabaDiskPerformanceLevel enum attribute to describe a disk's performance level
+type AlibabaDiskPerformanceLevel string
+
+// AlibabaDiskCatagory enum attribute to deescribe a disk's category
+type AlibabaDiskCategory string
+
+// AlibabaDiskEncryptionMode enum attribute to describe whether to enable or disable disk encryption
+type AlibabaDiskEncryptionMode string
+
+// AlibabaDiskPreservationPolicy enum attribute to describe whether to preserve or delete a disk upon instance removal
+type AlibabaDiskPreservationPolicy string
+
+const (
+	// DeleteWithInstance enum property to delete disk with instance deletion
+	DeleteWithInstance AlibabaDiskPreservationPolicy = "DeleteWithInstance"
+	// PreserveDisk enum property to determine disk preservation with instance deletion
+	PreserveDisk AlibabaDiskPreservationPolicy = "PreserveDisk"
+
+	// AlibabaDiskEncryptionEnabled enum property to enable disk encryption
+	AlibabaDiskEncryptionEnabled AlibabaDiskEncryptionMode = "encrypted"
+	// AlibabaDiskEncryptionDisabled enum property to disable disk encryption
+	AlibabaDiskEncryptionDisabled AlibabaDiskEncryptionMode = "disabled"
+
+	// AlibabaDiskPerformanceLevel0 enum property to set the level at PL0
+	PL0 AlibabaDiskPerformanceLevel = "PL0"
+	// AlibabaDiskPerformanceLevel1 enum property to set the level at PL1
+	PL1 AlibabaDiskPerformanceLevel = "PL1"
+	// AlibabaDiskPerformanceLevel2 enum property to set the level at PL2
+	PL2 AlibabaDiskPerformanceLevel = "PL2"
+	// AlibabaDiskPerformanceLevel3 enum property to set the level at PL3
+	PL3 AlibabaDiskPerformanceLevel = "PL3"
+
+	// AlibabaDiskCategoryUltraDisk enum proprty to set the category of disk to ultra disk
+	AlibabaDiskCatagoryUltraDisk AlibabaDiskCategory = "cloud_efficiency"
+	// AlibabaDiskCategorySSD enum proprty to set the category of disk to standard SSD
+	AlibabaDiskCatagorySSD AlibabaDiskCategory = "cloud_ssd"
+	// AlibabaDiskCategoryESSD enum proprty to set the category of disk to ESSD
+	AlibabaDiskCatagoryESSD AlibabaDiskCategory = "cloud_essd"
+	// AlibabaDiskCategoryBasic enum proprty to set the category of disk to basic
+	AlibabaDiskCatagoryBasic AlibabaDiskCategory = "cloud"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// AlibabaCloudMachineProviderConfig is the Schema for the alibabacloudmachineproviderconfig API
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+// +k8s:openapi-gen=true
+type AlibabaCloudMachineProviderConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// More detail about alibabacloud ECS
+	// https://www.alibabacloud.com/help/doc-detail/25499.htm?spm=a2c63.l28256.b99.727.496d7453jF7Moz
+
+	//The instance type of the instance.
+	InstanceType string `json:"instanceType"`
+
+	// The ID of the vpc
```

Comment:
describe what happens if `""`

Reply:
The VpcID is used to query and then match the machines. If this information is not included it is left empty. Here is a current machineset:
```
        providerSpec:
          value:
            apiVersion: alibabacloudmachineproviderconfig.openshift.io/v1beta1
            bandwidth: {}
            credentialsSecret:
              name: alibabacloud-credentials
            dedicatedHostId: ""
            imageId: m-0xi5sh1y90lsbo8ayl0a
            instanceType: ecs.g6.large
            kind: AlibabaCloudMachineProviderConfig
            metadata:
              creationTimestamp: null
            regionId: us-east-1
            resourceGroupId: rg-acfnxrgr6qtm3mq
            securityGroups:
            - tags:
              - Key: kubernetes.io/cluster/test-2qzw5
                Value: owned
              - Key: GISV
                Value: ocp
              - Key: sigs.k8s.io/cloud-provider-alibaba/origin
                Value: ocp
              - Key: Name
                Value: test-2qzw5-sg-worker
            subscription: {}
            systemDisk:
              category: cloud_essd
              size: 120
            tags:
            - Key: kubernetes.io/cluster/test-2qzw5
              Value: owned
            - Key: GISV
              Value: ocp
            - Key: sigs.k8s.io/cloud-provider-alibaba/origin
              Value: ocp
            userDataSecret:
              name: worker-user-data
            vSwitch:
              tags:
              - Key: kubernetes.io/cluster/test-2qzw5
                Value: owned
              - Key: GISV
                Value: ocp
              - Key: sigs.k8s.io/cloud-provider-alibaba/origin
                Value: ocp
              - Key: Name
                Value: test-2qzw5-vswitch-us-east-1b
            zoneId: us-east-1b
```
The VpcID is not included and therefore when queries are run they are excluded.
https://github.com/openshift/cluster-api-provider-alibaba/blob/main/pkg/actuators/machine/instances.go#L394
https://github.com/openshift/cluster-api-provider-alibaba/blob/main/pkg/actuators/machine/instances.go#L467

The end result is that the instance is found by the tags, region, and other metadata. This is just one piece of metadata that could be used to search for the instance.